subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConsumableDependenciesConfiguration.java

```diff
@@ -0,0 +1,111 @@
+ */
+public class DefaultConsumableDependenciesConfiguration extends DefaultConfiguration implements ConsumableConfiguration, DependenciesConfiguration {
+
+    public DefaultConsumableDependenciesConfiguration(
+        DomainObjectContext domainObjectContext,
+        String name,
+        ConfigurationsProvider configurationsProvider,
+        ConfigurationResolver resolver,
+        ListenerBroadcast<DependencyResolutionListener> dependencyResolutionListeners,
+        ProjectDependencyObservedListener dependencyObservedBroadcast,
+        DependencyMetaDataProvider metaDataProvider,
+        ComponentIdentifierFactory componentIdentifierFactory,
+        DependencyLockingProvider dependencyLockingProvider,
+        Factory<ResolutionStrategyInternal> resolutionStrategyFactory,
+        FileCollectionFactory fileCollectionFactory,
+        BuildOperationExecutor buildOperationExecutor,
+        Instantiator instantiator,
+        NotationParser<Object, ConfigurablePublishArtifact> artifactNotationParser,
+        NotationParser<Object, Capability> capabilityNotationParser,
+        ImmutableAttributesFactory attributesFactory,
+        RootComponentMetadataBuilder rootComponentMetadataBuilder,
+        ResolveExceptionContextualizer exceptionContextualizer,
+        UserCodeApplicationContext userCodeApplicationContext,
+        ProjectStateRegistry projectStateRegistry,
+        WorkerThreadRegistry workerThreadRegistry,
+        DomainObjectCollectionFactory domainObjectCollectionFactory,
+        CalculatedValueContainerFactory calculatedValueContainerFactory,
+        DefaultConfigurationFactory defaultConfigurationFactory,
+        TaskDependencyFactory taskDependencyFactory,
+        ConfigurationRole roleAtCreation,
+        boolean lockUsage
+    ) {
+        super(
+            domainObjectContext,
+            name,
+            configurationsProvider,
+            resolver,
+            dependencyResolutionListeners,
+            dependencyObservedBroadcast,
+            metaDataProvider,
+            componentIdentifierFactory,
+            dependencyLockingProvider,
+            resolutionStrategyFactory,
+            fileCollectionFactory,
+            buildOperationExecutor,
+            instantiator,
+            artifactNotationParser,
+            capabilityNotationParser,
+            attributesFactory,
+            rootComponentMetadataBuilder,
+            exceptionContextualizer,
+            userCodeApplicationContext,
+            projectStateRegistry,
+            workerThreadRegistry,
+            domainObjectCollectionFactory,
+            calculatedValueContainerFactory,
+            defaultConfigurationFactory,
+            taskDependencyFactory,
+            roleAtCreation,
+            lockUsage
+        );
+
+        assert roleAtCreation.isConsumable() && roleAtCreation.isDeclarable() && !roleAtCreation.isResolvable();
```

Comment:
This is a side-effect of the fact that all configurations (right now) have the same constructor. We _could_ hard-code the `roleAtCreation` for ResolvableConfiguration, ConsumableConfiguration and, DependenciesConfiguration, but the others need to have customizable roles since they can be instantiated in various combinations of deprecation. 

For example, we need to be able to create a legacy configuration that has no deprecations, and one which can satisfy `LEGACY_TO_RESOLVABLE_BUCKET`. 

I wouldn't necessarily be concerned about this assert. Its not that we're _not sure_ of what the state of the object is, but for simplicity given that all configuration types have the same constructor pattern, this both code-ifys the contract between role at creation and the class of the Configuration, and adds a sanity check to make sure we don't screw up anything to royally. 

With the API exposed to users, is is not possible to break this assertion. Any circumstance that would break this would be caught during testing. 

We _can_ change this, but it would just add a bunch of more duplicate boilerplate and IMO its not worth it.

Reply:
Agree we should avoid the extra boilerplate.  Maybe worth a comment explaining some of this and saying "don't worry".  I imagine a build script author shaking their head over seeing this, and also perhaps wondering why it isn't a `Precondition.checkState` or if it's meaningful or just forgotten debug code.

Reply:
These asserts have since been removed. This is no longer relevant